File: atlas/examples/typescript/packages/application/src/catalog/CatalogImportReport.ts
import type { CatalogItem } from '@atlas-example/domain';

/**
 * Reports the immutable result of one catalog import use case.
 */
export class CatalogImportReport {
  /**
   * Creates an import report.
   *
   * @param importedAt - Timestamp supplied by the application clock.
   * @param items - Items created and persisted by this import.
   */
  public constructor(
    public readonly importedAt: Date,
    public readonly items: readonly CatalogItem[],
  ) {}

  /**
   * Counts imported items without exposing a mutable collection.
   *
   * @returns Number of imported catalog items.
   */
  public get importedCount(): number {
    return this.items.length;
  }
}

File: atlas/examples/kotlin/application/src/main/kotlin/dev/atlas/example/application/catalog/CatalogImportReport.kt
package dev.atlas.example.application.catalog

import dev.atlas.example.domain.catalog.CatalogItem
import java.time.Instant

/**
 * Reports the immutable result of one catalog import use case.
 *
 * @property importedAt Timestamp supplied by the application clock.
 * @property items Items created and persisted by this import.
 */
data class CatalogImportReport(
    val importedAt: Instant,
    val items: List<CatalogItem>
) {
    /**
     * Counts imported items without exposing a mutable collection.
     */
    val importedCount: Int
        get() = items.size
}
